controllers/managedclustermodule_reconciler.go

```diff
@@ -0,0 +1,84 @@
+/*
+Copyright 2022.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	kmmv1beta1 "github.com/kubernetes-sigs/kernel-module-management/api/v1beta1"
+)
+
+// ManagedClusterModuleReconciler reconciles a ManagedClusterModule object
+type ManagedClusterModuleReconciler struct {
+	client.Client
+
+	Scheme *runtime.Scheme
+}
+
+//+kubebuilder:rbac:groups=kmm.sigs.k8s.io,resources=managedclustermodules,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=kmm.sigs.k8s.io,resources=managedclustermodules/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=kmm.sigs.k8s.io,resources=managedclustermodules/finalizers,verbs=update
+
+func NewManagedClusterModuleReconciler(
+	client client.Client) *ManagedClusterModuleReconciler {
+	return &ManagedClusterModuleReconciler{
+		Client: client,
+	}
+}
+
+func (r *ManagedClusterModuleReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	res := ctrl.Result{}
+
+	logger := log.FromContext(ctx)
+
+	mcm, err := r.getRequestedManagedClusterModule(ctx, req.NamespacedName)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			logger.Info("ManagedClusterModule deleted")
+			return res, nil
+		}
+
+		return res, fmt.Errorf("failed to get the requested %s KMMO CR: %w", req.NamespacedName, err)
+	}
+
+	logger.Info("Requested KMMO ManagedClusterModule", "namespace", mcm.Name, "name", mcm.Namespace)
+
+	return res, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *ManagedClusterModuleReconciler) SetupWithManager(mgr ctrl.Manager) error {
```

Comment:
I don't see any caller in `main.go`. I guess it is not implemented yet?!

Reply:
> @mresvanis this PR's reconciler is just a skeleton, right? Because i doesn't do much right now

That's correct, it's just a first skeleton (mostly what is scaffolded by `kubebuilder` and just a few minor details. I opened the PR (Draft) early, because it's gonna be too big (I believe) for only a final review.

Reply:
@ybettan this reconciler will be only included in the KMM-Hub version of KMM, i.e. either by adding a configuration flag or by building a dedicated KMM-Hub container image (this is a design discussion and it's still ongoing). As a result, we won't be enabling this reconciler in `main.go` in this PR. What do you think?

Reply:
I am wondering how it will look like. Correct me if I am wrong but the way I see it we have 2 options:
* runtime option: in that case we should have an `if` statement in `main` to start the controller in case some runtime argument were/weren't passed.
* buildtime option: in that case I wouldn't expect to compile the reconciler code in that image.

Am I missing something here?

Reply:
The runtime option is clear, you're not missing anything there :+1: 

Regarding the different builds option, currently all controllers belong to the same `controllers` package. Which means that since this package needs to be `import`ed in a `main.go` for a user to run one or more of those controllers, then all controllers will be compiled and get into the binary (as you already said). 

One option we have been thinking to alleviate this issue is to move each of the optional controllers to their own Golang package and import them explicitly in the respective `main`s (we would anyway have as many `main`s as binaries). That way all binaries would only contain what's absolutely needed.

But until we reach a consensus about `runtime` vs `build` options, I thought it best to include this reconciler along with the rest of the reconcilers. If we go for the `build` option, then we would move it to its own package and also implement the rest of the supporting details around this, e.g. create the additional `cmd/kmm-hub/main.go`.

Does this make sense?

Reply:
Yup. It make a lot of sense.

Seems like the build option is way too complicated for what we really need and I tend to go with the runtime option but this is indeed a conversation which is not tightly coupled with this PR.